PRCS/prc-2.md

```diff
@@ -0,0 +1,157 @@
+    /// Emits `Unlock` event.
+    /// @param tokens NFT tokens contract addresses
+    /// @param tokenIds NFT tokens identifiers
+    function withdraw(address[] memory tokens, uint256[] memory tokenIds) external;
+
+    /// @notice Changes `lockTime` variable that is used in `requestUnlock`.
+    /// @dev This function should be protected with appropriate access control mechanisms.
+    /// The new value should be checked against a sane upper limit constant, which if exceeded,
+    /// should cause a revert.
+    /// Emits `LockTimeUpdate` event.
+    /// @param newLockTime New lockTime value
+    function setLockTime(uint256 newLockTime) external;
+}
+```
+
+A Hololocker implementation MUST implement the IERC721Receiver interface to be able to receive ERC721 assets via `IERC721.safeTransferFrom`.
+It MUST initialize a lock similarly as in the `lock` function, and it MUST emit the `Lock` event
+
+```solidity
+interface IERC721Receiver {
+    /// @dev Whenever an {IERC721} `tokenId` token is transferred to this contract via {IERC721-safeTransferFrom}
+    ///  by `operator` from `from`, this function is called.
+    ///  It must return its Solidity selector to confirm the token transfer.
+    ///  If any other value is returned or the interface is not implemented by the recipient, the transfer will be
+    ///  reverted.
+    ///  The selector can be obtained in Solidity with `IERC721Receiver.onERC721Received.selector`.
+    ///  Note: the contract address is always the message sender.
+    /// @param _operator The address which called `safeTransferFrom` function
+    /// @param _from The address which previously owned the token
+    /// @param _tokenId The NFT identifier which is being transferred
+    /// @param _data Additional data with no specified format
+    /// @return `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`
+    ///  unless throwing
+    function onERC721Received(address _operator, address _from, uint256 _tokenId, bytes _data) external returns(bytes4);
+}
+```
+
+## Rationale
+
+The Hololocker contract is in its core a NFT staking contract. While there have been attempts at EIPs on this matter, such as [https://eips.ethereum.org/EIPS/eip-4987](EIP-4987), none gained enough traction and approval to be finalized.
+This proposal therefore proposes a standard for light-weight implementation of such mechanism.
+
+The main functions accept an array of token addresses and token IDs to be able to accomodate manipulations with multiple NFTs in one transaction, to save on transaction costs.
+
+### NFT locking UX
+
+Upon locking an NFT in an application that integrates the Hololocker, it might happen that users will get confused as to where did their NFTs go. An option to mint a sort of a "receipt NFT" in exchange for the locked NFT was thought of, but it does not sufficiently solve the inconvenience and would greatly increase the complexity and transactional costs. Therefore, this approach is not recommended.

```

Comment:
Extending this for ERC20 is not as straight forward as this makes it sounds. ERC20s have an "amount" so the events in their current form wouldn't support them. That's why the ERC4987 EIP we mentioned earlier also keeps track of an "amount" in the Hold/Release event logs

Reply:
True, there must be many changes in many places - the LockInfo struct, the events, the function parameters. But in my opinion, the changes are really quite straightforward and come down to a simple task - instead of using a token ID you use the amount of tokens. Yes, it results in a completely different interface (as is the case of IERC20Holder and IERC721Holder in EIP-4987 as well) which would make a case for a new PRC. I'll add this little information I just mentioned, but unless you'd want to describe the entire interface for ERC20 Hololocker in this PRC, I'm not sure what else to add.

Reply:
I think it's just about describing which parts of the interface should stay the same can change if people extend it to other token types